components/eamxx/src/share/util/eamxx_time_interpolation.cpp

```diff
@@ -0,0 +1,340 @@
+ * field managers.
+ * Conducts a simple linear interpolation between two points using
+ *        y* = w*y0 + (1-w)*y1
+ * where w = (t1-t*)/(t1-t0)
+ * Input:
+ *   time_in - A timestamp to interpolate onto.
+ * Output
+ *   A map of (field name) and (Field) pairs such that each field is the interpolated values and
+ *   the map makes it possible to quickly query individual outputs.
+ */
+void TimeInterpolation::perform_time_interpolation(const TimeStamp& time_in)
+{
+  // Declare the output map
+  std::map<std::string,Field> interpolated_fields;
+
+  // If data is handled by files we need to check that the timestamps are still relevant
+  if (m_file_data_triplets.size()>0) {
+    check_and_update_data(time_in);
+  }
+
+  // Gather weights for interpolation.  Note, timestamp differences are integers and we need a
+  // real defined weight.
+  const Real w_num = m_time1 - time_in;
+  const Real w_den = m_time1 - m_time0;
+  const Real weight = w_num/w_den;
+
+  // Cycle through all stored fields and conduct the time interpolation
+  for (auto name : m_field_names)
+  {
+    const auto& field0   = m_fm_time0->get_field(name);
+    const auto& field1   = m_fm_time1->get_field(name);
+          auto field_out = m_interp_fields.at(name);
+    field_out.deep_copy(field0);
+    field_out.update(field1,1.0-weight,weight);
+  }
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function which registers a field in the local field managers.
+ * Input:
+ *   field_in - Is a field with the appropriate dimensions and metadata to match the interpolation
+ * Output:
+ *   None
+ */
+void TimeInterpolation::add_field(Field& field_in, const bool deep)
+{
+  // First check that we haven't already added a field with the same name.
+  const auto name = field_in.name();
+  EKAT_REQUIRE_MSG(!m_fm_time0->has_field(name) and !m_fm_time1->has_field(name),
+		  "Error!! TimeInterpolation:add_field, field + " << name << " has already been added." << "\n");
+
+  // Clone the field for each field manager to get all the metadata correct.
+  auto field0 = field_in.clone();
+  auto field1 = field_in.clone();
+  m_fm_time0->add_field(field0);
+  m_fm_time1->add_field(field1);
+  if (deep) {
+    // Then we want to store the actual field_in and override it when interpolating
+    m_interp_fields.emplace(name,field_in);
+  } else {
+    // We want to store a copy of the field but not ovveride
+    auto field_out = field_in.clone();
+    m_interp_fields.emplace(name,field_out);
+  }
+  m_field_names.push_back(name);
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function to shift all data from time1 to time0, update timestamp for time0
+ */
+void TimeInterpolation::shift_data()
+{
+  for (auto name : m_field_names)
+  {
+    auto field0 = m_fm_time0->get_field(name);
+    auto field1 = m_fm_time1->get_field(name);
+    field0.deep_copy(field1);
```

Comment:
Using std::swap should be the way to go here. We should find a way to make it work.